protocol/swap/status_manager.go

```diff
@@ -0,0 +1,63 @@
+package swap
+
+import (
+	"sync"
+
+	"github.com/athanorlabs/atomic-swap/common/types"
+)
+
+// statusManager provides lookup for the status channels. Status channels are
+// ephemeral between runs of swapd.
+type statusManager struct {
+	mu             sync.Mutex
+	statusChannels map[types.Hash]chan Status
+}
+
+func newStatusManager() *statusManager {
+	return &statusManager{
+		mu:             sync.Mutex{},
+		statusChannels: make(map[types.Hash]chan Status),
+	}
+}
+
+// getStatusChan returns any existing status channel or a new status channel for
+// reading or writing.
+func (sm *statusManager) getStatusChan(offerID types.Hash) chan Status {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	_, ok := sm.statusChannels[offerID]
+	if !ok {
+		sm.statusChannels[offerID] = newStatusChannel()
+	}
+
+	return sm.statusChannels[offerID]
+}
+
+// GetStatusChan returns any existing status channel or a new status channel for
+// reading only.
+func (sm *statusManager) GetStatusChan(offerID types.Hash) <-chan Status {
+	return sm.getStatusChan(offerID)
+}
+
+// DeleteStatusChan deletes any status channel associated with the offer ID.
+func (sm *statusManager) DeleteStatusChan(offerID types.Hash) {
+	sm.mu.Lock()
+	defer sm.mu.Unlock()
+
+	delete(sm.statusChannels, offerID)
+}
+
+// PushNewStatus adds a new status to the offer ID's channel
+func (sm *statusManager) PushNewStatus(offerID types.Hash, status types.Status) {
```

Comment:
we should remove `info.SetStatus` and make a function on the swapManager that updates the status that also calls this function

Reply:
See what I implemented and if you like it. I was originally thinking along the lines of what you wrote above, but I decided that this manager should only deal with maintaining the map of the status channel (which has a different lifecycle than the `Info` struct). Locking and changing the `Info` struct is handled by member functions that are invoked by the maker/taker `swapStage` go-processes.